### PR TITLE
[Storage] Bucket lock samples

### PR DESCRIPTION
**Pull request by frankyn**
### Samples
- [x]  Tested samples
- [x]  Define a retention policy
- [x]  Remove a retention policy if not locked
- [x]  Lock a retention policy
- [x]  Set temporary hold
- [x]  Release temporary hold
- [x]  Set event-based hold
- [x]  Release event-based hold
- [x]  Set default event-based hold
- [x]  Release default event-based hold
- [x]  View default event-based hold
- [x]  View bucket retention policy metadata
- [x]  View metadata for object
- [x]  Updated README.md
- [x]  Updated CLI
- [x] enable feature for test projects.

### CODE REVIEW

**Comment by frankyn** [comment]
Summoning @yoshi-ruby to review my PR. Anyone have a moment to take a look?

**Comment by frankyn** [comment]
Summoning @GoogleCloudPlatform/yoshi-ruby to review my PR. Anyone have a moment to take a look?

**Comment by quartzmo** [comment]
@frankyn I'll review this first thing tomorrow morning!

File: storage/buckets.rb [hunk 1]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
 
+def set_retention_policy project_id:, bucket_name:, retention_period:
+  # [START storage_set_retention_policy]
+  # project_id       = "Your Google Cloud project ID"
+  # bucket_name      = "Name of your Google Cloud Storage bucket"
+  # retention_period = "Object retention period defined in seconds"
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
```
Comment (blowmage) [inline storage/buckets.rb]: We are encouraging use of the `project_id` argument over the `project` argument. 

```
storage = Google::Cloud::Storage.new project_id: project_id
```

Comment (frankyn) [reply storage/buckets.rb]: acking

**Comment by frankyn** [comment]
thanks both!

File: storage/buckets.rb (hunk 1 above)
Comment (blowmage) [reply storage/buckets.rb]: We also encourage use of the `credentials` argument over `keyfile`, but I don't know if you are using that.

File: storage/buckets.rb [hunk 2]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  # bucket_name = "Name of your Google Cloud Storage bucket"
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+
+  # Warning: Once a retention policy is locked it cannot be unlocked
+  # and retention period can only be increased.
+  # Uses bucket.metageneration as a precondition
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

Comment (quartzmo) [reply storage/buckets.rb]: If you mean the public Ruby API, bucket.metageneration should be `Bucket#metageneration`.

File: storage/buckets.rb [hunk 3]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  bucket.lock_retention_period!
+
+  puts "Retention policy for #{bucket_name} is now locked"
+  puts "Retention policy effective as of #{bucket.retention_effective_at}"
+  # [END storage_lock_retention_policy]
```
Comment (quartzmo) [inline storage/buckets.rb]: Sentences above are both missing periods.

File: storage/buckets.rb [hunk 4]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+
+  if !bucket.retention_policy_locked?
+    bucket.retention_period = nil
+    puts "Retention period for #{bucket_name} has been removed."
+  else
+    puts "Policy is locked and retention periodcan't be removed."
```
Comment (quartzmo) [inline storage/buckets.rb]: Add space: "periodcan't"

File: storage/buckets.rb [hunk 5]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  # bucket_name = "Name of your Google Cloud Storage bucket"
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+
+  if !bucket.retention_policy_locked?
+    bucket.retention_period = nil
+    puts "Retention period for #{bucket_name} has been removed."
```
Comment (quartzmo) [inline storage/buckets.rb]: Do you want to say "Retention policy for..." instead? You can if you want, per the docs:

> Passing `nil` will remove the retention policy from the bucket if it exists, unless the policy is locked.

File: storage/buckets.rb [hunk 6]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+
+  bucket.update do |b|
+    b.default_event_based_hold = true
+  end
+
+  puts "Default event based hold was enabled for #{bucket_name}"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 7]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+
+  bucket.update do |b|
+    b.default_event_based_hold = false
+  end
+
+  puts "Default event based hold was disabled for #{bucket_name}"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 8]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  # project_id  = "Your Google Cloud project ID"
+  # bucket_name = "Name of your Google Cloud Storage bucket"
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+
+  if bucket.default_event_based_hold?
+    puts "Default event based hold is enabled for {bucket_name}"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 9]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  else
+    puts "Default event based hold is not enabled for {bucket_name}"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/files.rb [hunk 10]
```diff
@@ -382,6 +386,78 @@ def generate_signed_url project_id:, bucket_name:, file_name:
...
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+  file    = bucket.file file_name
+
+  file.set_event_based_hold!
+
+  puts "Event based hold was set for #{file_name}"
```
Comment (quartzmo) [inline storage/files.rb]: This sentence is missing a period at the end.

File: storage/files.rb [hunk 11]
```diff
@@ -382,6 +386,78 @@ def generate_signed_url project_id:, bucket_name:, file_name:
...
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+  file    = bucket.file file_name
+
+  file.release_event_based_hold!
+
+  puts "Event based hold was released for #{file_name}"
```
Comment (quartzmo) [inline storage/files.rb]: This sentence is missing a period at the end.

File: storage/files.rb [hunk 12]
```diff
@@ -382,6 +386,78 @@ def generate_signed_url project_id:, bucket_name:, file_name:
...
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+  file    = bucket.file file_name
+
+  file.set_temporary_hold!
+
+  puts "Temporary hold was set for #{file_name}"
```
Comment (quartzmo) [inline storage/files.rb]: This sentence is missing a period at the end.

File: storage/files.rb [hunk 13]
```diff
@@ -382,6 +386,78 @@ def generate_signed_url project_id:, bucket_name:, file_name:
...
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project: project_id
+  bucket  = storage.bucket bucket_name
+  file    = bucket.file file_name
+
+  file.release_temporary_hold!
+
+  puts "Temporary hold was release for #{file_name}"
```
Comment (quartzmo) [inline storage/files.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 14]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+
+  # Warning: Once a retention policy is locked it cannot be unlocked
+  # and retention period can only be increased.
+  # Uses bucket.metageneration as a precondition
+  bucket.lock_retention_policy!
+
+  puts "Retention policy for #{bucket_name} is now locked"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 15]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+
+  bucket.update do |b|
+    b.default_event_based_hold = true
+  end
+
+  puts "Default event-based hold was enabled for #{bucket_name}"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 16]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+
+  bucket.update do |b|
+    b.default_event_based_hold = false
+  end
+
+  puts "Default event-based hold was disabled for #{bucket_name}"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 17]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  # project_id  = "Your Google Cloud project ID"
+  # bucket_name = "Name of your Google Cloud Storage bucket"
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+
+  if bucket.default_event_based_hold?
+    puts "Default event-based hold is enabled for {bucket_name}"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 18]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  else
+    puts "Default event-based hold is not enabled for {bucket_name}"
```
Comment (quartzmo) [inline storage/buckets.rb]: This sentence is missing a period at the end.

File: storage/files.rb [hunk 19]
```diff
@@ -382,6 +386,78 @@ def generate_signed_url project_id:, bucket_name:, file_name:
...
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+  file    = bucket.file file_name
+
+  file.set_event_based_hold!
+
+  puts "Event-based hold was set for #{file_name}"
```
Comment (quartzmo) [inline storage/files.rb]: This sentence is missing a period at the end.

File: storage/files.rb [hunk 20]
```diff
@@ -382,6 +386,78 @@ def generate_signed_url project_id:, bucket_name:, file_name:
...
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+  file    = bucket.file file_name
+
+  file.release_event_based_hold!
+
+  puts "Event-based hold was released for #{file_name}"
```
Comment (quartzmo) [inline storage/files.rb]: This sentence is missing a period at the end.

File: storage/files.rb [hunk 21]
```diff
@@ -382,6 +386,78 @@ def generate_signed_url project_id:, bucket_name:, file_name:
...
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+  file    = bucket.file file_name
+
+  file.release_temporary_hold!
+
+  puts "Temporary hold was released for #{file_name}"
```
Comment (quartzmo) [inline storage/files.rb]: This sentence is missing a period at the end.

File: storage/buckets.rb [hunk 22]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  # retention_period = "Object retention period defined in seconds"
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+
+  bucket.retention_period = retention_period
+
+  puts "Retention period for #{bucket_name} is now #{retention_period} seconds"
```
Comment (quartzmo) [inline storage/buckets.rb]: I would have more confidence in this output if it accessed the bucket fields: `#{bucket.name} is now #{bucket.retention_period}`

File: storage/buckets.rb (hunk 16 above)
Comment (quartzmo) [reply storage/buckets.rb]: What about: "Default event-based hold is now #{bucket.default_event_based_hold?} for #{bucket_name}."

File: storage/files.rb (hunk 12 above)
Comment (frankyn) [reply storage/files.rb]: Wanted to keep it contextual to the sample.

**Comment by frankyn** [comment]
Responded to feedback, thanks @quartzmo!

File: storage/buckets.rb [hunk 23]
```diff
@@ -199,6 +199,133 @@ def delete_bucket project_id:, bucket_name:
...
+  # retention_period = "Object retention period defined in seconds"
+
+  require "google/cloud/storage"
+
+  storage = Google::Cloud::Storage.new project_id: project_id
+  bucket  = storage.bucket bucket_name
+
+  bucket.retention_period = retention_period
+
+  puts "Retention period for #{bucket_name} is now #{bucket.retention_period} seconds"
```
Comment (quartzmo) [inline storage/buckets.rb]: Missing period at end of sentence.

Comment (frankyn) [reply storage/buckets.rb]: Whoops, done!

**Comment by frankyn** [comment]
Thank you @quartzmo and @blowmage! Super appreciate it! Now I need to figure out what happened with Kokoro CI, looks like a permission issue :/

**Comment by tbpg** [comment]
Kokoro might take a bit of time to get back to normal. So, if tests are passing locally, I would submit.

**Comment by frankyn** [comment]
Thanks @tbpg, do you know if this is related to the new trampoline release?

**Comment by frankyn** [comment]
Also, I'm blocked until the library is released to rubygems.

**Comment by tbpg** [comment]
> do you know if this is related to the new trampoline release?
I think so.

**Comment by frankyn** [comment]
Thanks @tbpg. I will merge next week, hopefully it will be rolled back by then.

**Comment by frankyn** [comment]
Thanks y'all. Merging now.

**Pull request merged**